### PR TITLE
Complete only missing fields in pats

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
A follow-up for https://github.com/rust-analyzer/rust-analyzer/pull/3694

Same name vs string [issue](https://github.com/rust-analyzer/rust-analyzer/pull/3694#discussion_r396986819) persists here, now I'm able to obtain `ast::Name`, but I see no way to convert it into `hir::Name` or vice versa.

### CODE REVIEW

**Comment by flodiebold** [comment]
I think a method on `Semantics` that takes an `ast::Name` and returns a `hir::Name` might make sense. It would be able to do hygiene properly when that's necessary/implemented.

**Comment by flodiebold** [comment]
On the other hand, going to the `hir::Pat` for the whole pattern first might actually be the better approach here.

**Comment by Veetaha** [comment]
I had thoughts on making `hir::Pat`, but have quite approached it due to the lack of time.
I described my concerns [in here](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/hover/near/192099783). I think we should make it smtg like:
```rust
struct ra_hir::Pat {
    id: Idx<ra_hir_def::Pat>,
    parent: ra_hir_def::DefWithBodyId,
}
```

**Comment by SomeoneToIgnore** [comment]
>  going to the hir::Pat for the whole pattern first might actually be the better approach here.

Hm, I see that `ctx` has `sema: Semantics` and it has a way to get the `SourceAnalyzer` which has a way to get a `PatId` for the `ast::RecordPat` given, same as it has the `Body` to derive the `Pat` based on that id.
All the methods and fields I need are private, but it looks like I can hack around things and get the `Pat` for the given `ast:RecordPat`. 
Or is there a simpler way?

Next, I have to match the `Pat` contents and descend down to the `Path::Bind` to extract the name and add it to the vec of an existing name.

Did I get the idea right?
It looks doable, but what is the actual benefit compared to the ast approach?
If I got that right, the [issue](https://github.com/rust-analyzer/rust-analyzer/pull/3694#discussion_r396986819) I've mentioned in the PR message was related to the fact that we do unnecessary allocations, yet I'm not sure that we avoid them by going this path.

**Comment by Veetaha** [comment]
@SomeoneToIgnore I may be wrong, but we shouldn't expose `ra_hir_def::Pat` to ~~the wild~~ `ra_ide`. This enum contains `PatId` inside of some of its variants and I remember matklad sayin' that the `ra_ide` crate should not interface with anything `id`-based, so that this is hidden by the `ra_hir` layer ([citation from him](https://github.com/rust-analyzer/rust-analyzer/blob/6f0d8db529478ce41b429f06708fa600a97c2151/crates/ra_hir/src/lib.rs#L12-L18)).

**Comment by SomeoneToIgnore** [comment]
They are not available in `ra_ide` currently, good point, thanks.

> going to the hir::Pat for the whole pattern first might actually be the better approach here.

So apparently I've understood it wrong and need clarification :)

File: crates/ra_ide/src/completion/complete_record_pattern.rs [hunk 1]
```diff
@@ -10,8 +11,24 @@ pub(super) fn complete_record_pattern(acc: &mut Completions, ctx: &CompletionCon
         _ => return,
     };
 
+    let already_present_names: Vec<SmolStr> = ctx
+        .record_lit_pat
+        .as_ref()
+        .and_then(|record_pat| record_pat.record_field_pat_list())
+        .map(|pat_list| pat_list.bind_pats())
```
Comment (flodiebold) [inline crates/ra_ide/src/completion/complete_record_pattern.rs]: you also need to look at `record_field_pats`. `bind_pats` are only the field shorthand patterns, you're currently missing any `foo1: something`.

**Comment by matklad** [comment]
>I think a method on Semantics that takes an ast::Name and returns a hir::Name might make sense. It would be able to do hygiene properly when that's necessary/implemented.

I think for this specifically, we should have something more high-level

```rust
impl Semantics {
    fn missing_fields(&self, src: RecordLiteral | CallExpr | TupleStructPat | RecordPat) -> Vec<StructField>
}
```

In general, I think we should strive for broad, but shallow API, a-la [BindingContext](https://github.com/JetBrains/kotlin/blob/61ad32f012789ef95376ba30812c32e779c4806b/compiler/frontend/src/org/jetbrains/kotlin/resolve/BindingContext.java#L96-L108).

**Comment by matklad** [comment]
Specific action items for this PR are probably:

* add initial implementation and tests using the current approach
* based on this experience, move the logic from here to and `complete_record_literal` to semantics (via a couple of functions which return a Vec of missing `StructField`).

**Comment by flodiebold** [comment]
> I think for this specifically, we should have something more high-level
> 
> ```rust
> impl Semantics {
>     fn missing_fields(&self, src: RecordLiteral | CallExpr | TupleStructPat | RecordPat) -> Vec<StructField>
> }
> ```
> 
> In general, I think we should strive for broad, but shallow API, a-la [BindingContext](https://github.com/JetBrains/kotlin/blob/61ad32f012789ef95376ba30812c32e779c4806b/compiler/frontend/src/org/jetbrains/kotlin/resolve/BindingContext.java#L96-L108).

Hmm to me, 'missing fields' seems like an IDE-level concern and doesn't really fit into my understanding of `Semantics` :thinking: (It's not a 'diagnostic' either since we would want to consider all fields 'missing' in `Foo { .. }`.)

**Comment by matklad** [comment]
Yeah, it's true that it is primarily for IDEs (although I expect there will be some impl intersection with relevant diagnostics). But in my mind it's not really about IDE vs compiler. Like we only use Semantics from an IDE anyway, and the whole `source_do_def` machinery is IDE-specific.

Rather, it's about the level of abstraction. Stuff below semantics works with relative ids, stuff above works with OO-ish code-model, and Semantics is a facade. So, I wouldn't worry about `Semantics` turning into a kitchen sink: as semantics mehtods mostly return result directly and do not compos with each other, adding or removing `Semantic`s method should be cheap. OTOH, if we expose more of the internals (like ids), than we'll get the API where you need to compose several methods to get the desired result, and in this situation changing a single method would be costly. 

Yeah, I sound like I know what I am talking about, but I actually don't, it's just my gut feeling that it's better to expose sepecific rather than general APIs here.

**Comment by SomeoneToIgnore** [comment]
And I just wanted to submit a small useful PR 😀
Thank you for mapping the path, I'll try to implement it and see what comes out of it.

**Comment by matklad** [comment]
To clarify, doing just 

>add initial implementation and tests using the current approach

is totally fine, I can then look into refactoring things :)

**Comment by SomeoneToIgnore** [comment]
Just to clarify more: what is the **current** approach?
The `ast` one or the https://github.com/rust-analyzer/rust-analyzer/pull/3779#issuecomment-606293329 one?

**Comment by Veetaha** [comment]
That BindingContext thing requires quite some digging...

**Comment by matklad** [comment]
The ast one, or, really, whichever simpler that works in the common case :)

**Comment by SomeoneToIgnore** [comment]
I've fiddled with the code and eventually decided to unite both files into one: whenever I extract the common logic, the rest of it in both files becomes so small and similar, that it feels like the right way. 
Wdyt?

Few nonrelated findings along the way:

* we don't create a runner for a top-level test mod now, so we cannot run all tests at once, we might want to fix that

* cannot autoimport `ra_syntax::ast::NameOwner` trait, I'll investigate this further

File: crates/ra_ide/src/completion/complete_record_pattern.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/ra_ide/src/completion/complete_record_pattern.rs]: Fixed and amended the test, thanks.

**Comment by matklad** [comment]
Yeah, I do think merging the two makes sense. 

>we don't create a runner for a top-level test mod now, so we cannot run all tests at once, we might want to fix that

Yes please, this have been bugging me for a long time

bors r+

**Comment by SomeoneToIgnore** [comment]
> cannot autoimport ra_syntax::ast::NameOwner trait, I'll investigate this further

That is explained by the fact that all closure parameters are `{unknown}` currently:

```rust
.filter_map(|fild_pat| fild_pat.name())
.chain(pat_list.bind_pats().filter_map(|bind_pat| bind_pat.name()))
```

hence the trait checks in `iterate_method_candidates` do not pass.

**Comment by flodiebold** [comment]
> That is explained by the fact that all closure parameters are `{unknown}` currently

Not *all* of them, just *many* :grimacing:

**Pull request merged**